### PR TITLE
Fix package build and refactor tsconfig to be less complex and breakable

### LINKED ISSUE #2907: TypeScript `defineComponent` type inference crash
## Description

TypeScript's type inference fails on our component files with the following error:

```
The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.
```

## Reproduction

1.  Open `VInputModal.vue` in VSCode (other editors too, I presume, but make sure to use the `vue-tsc` server, not something built-in, whether VSCode's or WebStorm) and see the error on the default exported component.

## Additional context

Marked as high and added to the todos because I think this might actually hinder our type checking. Something deeper may be going on that is causing our type checking to incorrectly pass despite these errors occurring in the file.

### LINKED ISSUE #2965: ESLint plugin package is broken on `main`
## Description

The ESLint plugin package is broken on `main` with unit tests failing. This breakage was likely introduced in PR #2908, similar to #2957.

## Reproduction

1. Check out repo.
2. Run unit tests for `@openverse/eslint-plugin`.

Alternatively, see [this CI + CD failure](https://github.com/WordPress/openverse/actions/runs/6055992747/job/16435584020#step:4:58).

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Fixes #2965  by @dhruvkb 
Fixes #2907 by @sarayourfriend (cc @obulat who will be happy about this)

## Description

When upgrading the ESLint package, we missed that there was a breakage. This breakage was caused by a pretty significant breaking change in the package we missed (me in implementation and others in review): https://github.com/typescript-eslint/typescript-eslint/issues/7284

To fix the issue, we needed to change the module resolution strategy used by TypeScript. However, because Nuxt has _very specific_ TypeScript requirements, making that change in the monolith configuration we implemented previously was impossible. Nuxt needs specific requirements that now clashed with the specific requirements of another project. `vue-tsc` doesn't support composite projects that would simplify this. This is a frustrating mess.

To fix it, I decided it would be easiest to isolate Nuxt's TypeScript configuration from everything else. Therefore, I extracted everything that _can_ be shared (basically, TypeScript's "linting" style checks, unused variables, etc) into `tsconfig.base.json` and then created individual `tsconfig.json`s in the `eslint-plugin` package and the `frontend` directory. Now both of them have their own, fully isolated TypeScript configurations. This essentially prevented the need to make any changes to the frontend to accommodate another package that relied on similarly opinionated dependencies.

And that's essentially what is at the core of this. We have multiple opinionated dependencies that cannot accommodate each other. As is so often the case, the easiest thing was to just separate the two and never have them speak again :slightly_smiling_face: .

In addition to that, the changes in the TypeScript ESLint dependency also broke `jest`'s built-in TypeScript compatibility. The ESLint package's maintainers suggested using `swc` as a drop-in replacement and indeed, it's very fast and required nothing more than installing two new packages and adding a single-line Jest configuration to the `eslint-plugin` package. Here again is something other packages may need to depend on.

To summarise the changes in this PR:

- Split the monolithic TypeScript configuration to a per-project approach. Now each package has its own `types` script in package.json that `pre-commit` will run recursively (mirroring how unit tests are handled).
   - A knock-on effect of this is that Storybook now thinks it's running a TypeScript project and insists on `ts-node` being available. I tried to disable this behaviour (not sure why it wasn't a problem pre-monorepo, because `tsconfig.json` existed in scope of storybook then as well) but it was easier to just add `ts-node` as a dev dependency.
   - Another knock-on effect is the ability to remove the types dependencies from the root package.json. Now packages just declare whatever `@types` et co packages they need rather than the root package.json needing to concern itself with that.
- Update `eslintrc.js` to reference the correct TypeScript configuration for each project so that path resolution for the `import` ruleset works using the TypeScript config.
- Replace `jest`'s built-in TypeScript configuration with `swc` for the `eslint-plugin` package to work around the `@typescript-eslint` suite of packages' opinionated approach to module resolution and `jest`'s lack of support for that.
   - `swc` was cool enough that I spent some time looking into the configuration of it and discovered that it could very easily replace babel. It's faster and the configuration is a bit less obscure (doesn't require additional plugins for ESLint compat).
- Upgrade the `@typescript-eslint` packages in the `eslint-plugin` as well and fix type errors introduced by the upgraded packages. This is another good indication that we should consider moving our entire eslint configuration into the plugin, so we don't have multiple places to upgrade eslint related dependencies.
- Remove `declaration: true` from all our TypeScript configurations. This gets rid of all the "this probably isn't portable" errors everywhere we saw them before and therefore fixes #2907.

## Testing Instructions

Checkout the branch and run `pnpm install` to get all the updated and new dependencies. Run `pnpm run -r types` and `pnpm run -r test:unit` and confirm everything passes. Make changes that would break TypeScript (add `export const s: number = 'asdf'` in frontend or packages and confirm `pnpm run -r types` raises the type error). CI should also confirm this. Run `just p frontend dev` and confirm the dev build works. Run `just p frontend prod` and confirm that works as well. Run `just frontend/build-img && just frontend/run-img` and confirm that works as well. **CI will _not_ test that the image actually works, just the build, as far as I understand it, so this must be tested locally. Staging would break immediately if it didn't work anyway.**. Rely on CI to confirm that the Playwright tests run, but do confirm they run and that the container didn't incorrectly exit early with a false `0` code.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by sarayourfriend** [comment]
Happy side-effect of this: TypeScript in the Nuxt project has never worked better for me than now. Even go-to-definition works perfectly with Vue components. It's _never_ worked for me before.

**Comment by sarayourfriend** [comment]
This is super frustrating. I fixed the tests in packages, but upgrading jest breaks the tests in the frontend because the transforms no longer work for dependencies. These opinionated dependencies are :face_exhaling:

**Comment by sarayourfriend** [comment]
It turns out that the frontend _needs_ Jest 26. On the other hand, the `@typescript-eslint` packages, _need_ Jest 29 (to get updates to `jest-resolve` that accommodate the way they organise their packages). 

That means we need to pin very specific versions of Jest dependencies in the packages. Or maybe in the frontend, which seems better, so that we're only worried about keeping the frontend at 26, rather than forcing anything else to go higher. It's an unfortunate and exhausting mess.

**Review by obulat — APPROVED** [review]
Wow, you are definitely right, I'm really happy about this PR 🎉 

I tried going to the definitions both in VS Code and in WebStorm, checked the `InstanceType<typeof VSomeComponent>` refs - everything works, and there are no problems in VS Code and no more red squigglies!
Tried committing changes (linting fails when there are errors) and pushing them upstream (no more local failures 🎉 ).

VS Code complains about tsconfig in `packages/eslint-plugin`, saying that "Option 'module' must be set to 'NodeNext' when option 'moduleResolution' is set to 'NodeNext'.ts" 

I played with Nuxt 3 yesterday, and noticed that the previous tsconfig setup didn't work with it, because the generated tsconfig's paths and some other properties would be overwritten. The changes in this PR will greately simplify ts configuration with Nuxt 3.

File: .eslintrc.js [hunk 1]
```diff
@@ -188,6 +188,27 @@ module.exports = {
         "tsdoc/syntax": "error",
       },
     },
+    {
+      files: ["frontend/**"],
+      settings: {
+        "import/resolver": {
+          typescript: {
+            project: "frontend/tsconfig.json",
+            extensions: [".js", ".ts", ".vue", ".png"],
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply .eslintrc.js:197]: I think that's right, yes. It was in the eslintrc before. We must not be importing SVGs into the test files :sweat_smile:

**Comment by sarayourfriend** [comment]
Thanks for the review, Olga. I'll make those changes now.

**Review by dhruvkb — APPROVED** [review]
LGTM! Also happy to see `swc` in our toolchain.

**Pull request merged**